src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,230 @@
+        if (index == size) {
+            linkedLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
+    }
+
+    void linkBefore(T e, Node<T> succ) {
+        final Node<T> pred = succ.previous;
+        final Node<T> newNode = new Node<>(pred, e, succ);
+        succ.previous = newNode;
+
+        if (pred == null) {
+            firstNode = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        addAll(size, list);
+    }
+
+    public void addAll(int index, Collection<? extends T> c) {
+        checkPositionIndex(index);
+
+        Object[] a = c.toArray();
+        int numNew = a.length;
+        if (numNew == 0) {
+            return;
+        }
+
+        Node<T> predecessor;
+        Node<T> successor;
+        if (index == size) {
+            successor = null;
+            predecessor = lastNode;
+        } else {
+            successor = node(index);
+            predecessor = successor.previous;
+        }
+
+        for (Object o : a) {
+            @SuppressWarnings("unchecked") T e = (T) o;
+            Node<T> newNode = new Node<>(predecessor, e, null);
+            if (predecessor == null) {
+                firstNode = newNode;
+            } else {
+                predecessor.next = newNode;
+            }
+            predecessor = newNode;
+        }
+
+        if (successor == null) {
+            lastNode = predecessor;
+        } else {
+            predecessor.next = successor;
+            successor.previous = predecessor;
+        }
+        size += numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkedElementIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkedElementIndex(index);
+        Node<T> x = node(index);
+        T oldValue = x.item;
+        x.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkedElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = firstNode; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = firstNode; x != null; x = x.next) {
+                if (object.equals(x.item)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> x) {
```

Comment:
Create a separate method `unlink(Node node)`. This logic will be used in methods when we remove from the list by index and by value.